src/js/binaryen.js-post.js

```diff
@@ -4414,54 +4433,147 @@ Module['Function'] = (() => {
   return Function;
 })();
 
+export { BynFunction as Function };
+
 // Additional customizations
 
-Module['exit'] = function(status) {
+MODULE['exit'] = function(status) {
   // Instead of exiting silently on errors, always show an error with
   // a stack trace, for debuggability.
   if (status != 0) throw new Error('exiting due to error: ' + status);
 };
 
-// Indicates if Binaryen has been loaded and is ready
-Module['isReady'] = runtimeInitialized;
-
 // Provide a mechanism to tell when the module is ready
 //
-// if (!binaryen.isReady) await binaryen.ready;
+// await binaryen.ready;
 // ...
 //
 let pendingPromises = [];
 let initializeError = null;
 
-Object.defineProperty(Module, 'ready', {
-  get() {
-    return new Promise((resolve, reject) => {
-      if (initializeError) {
-        reject(initializeError);
-      } else if (runtimeInitialized) {
-        resolve(Module);
-      } else {
-        pendingPromises.push({ resolve, reject });
-      }
-    });
+export const ready = new Promise((resolve, reject) => {
+  if (initializeError) {
+    reject(initializeError);
+  } else if (runtimeInitialized) {
+    resolve(MODULE);
+  } else {
+    pendingPromises.push({ resolve, reject });
   }
 });
 
 // Intercept the onRuntimeInitialized hook if necessary
 if (runtimeInitialized) {
   initializeConstants();
 } else {
-  Module['onRuntimeInitialized'] = (super_ => () => {
+  MODULE['onRuntimeInitialized'] = (super_ => () => {
     try {
       initializeConstants();
       if (super_) super_();
-      Module['isReady'] = true;
-      pendingPromises.forEach(p => { p.resolve(Module) });
+      pendingPromises.forEach(p => { p.resolve(MODULE) });
     } catch (e) {
       initializeError = e;
       pendingPromises.forEach(p => { p.reject(e) });
     } finally {
       pendingPromises = [];
     }
-  })(Module['onRuntimeInitialized']);
+  })(MODULE['onRuntimeInitialized']);
 }
+
+// Default export one gets upon either
+// * `import binaryen from "binaryen"` or
+// * `const binaryen = require("binaryen")`.
+export default {
```

Comment:
The repeated code here is somewhat typical for hybrid modules, in that it ensures that the module works ergonomically the same when `require`d. When transpiled to CommonJS, this becomes `module.exports`.

Reply:
Most (all?) of the functions here seem to be duplications of inline `export` statements. Why do you need them both as inline exports and as default object? Ideally you'd only keep the former, and transpilers can handle their conversion to CommonJS fine. It's also preferable, because named exports are tree-shakeable by other tooling, whereas with default object you lose this benefit of ESM.